source/common/stats/thread_local_store.cc

```diff
@@ -263,22 +264,43 @@ ThreadLocalStoreImpl::ScopeImpl::~ScopeImpl() {
   prefix_.free(symbolTable());
 }
 
-// Manages the truncation and tag-extraction of stat names. Tag extraction occurs
+// Helper for managing the potential truncation and extraction of tags from the metric names and
+// converting them to StatName. Making the tag extraction optional within this class simplifies the
+// RAII ergonomics (as opposed to making the construction of this object conditional).
+//
+// The StatNameTagVector returned by this object will be valid as long as this object is in scope
+// and the provided stat_name_tags are valid.
+//
+// When tag extraction is not done, this class is just a passthrough for the provided name/tags.
+//
+// Optionally manages the truncation and tag-extraction of stat names. Tag extraction occurs
 // on the original, untruncated name so the extraction can complete properly,
 // even if the tag values are partially truncated.
-class TagExtraction {
+class StatNameTagHelper {
 public:
-  TagExtraction(ThreadLocalStoreImpl& tls, StatName name) {
-    tls.symbolTable().callWithStringView(name, [this, &tls](absl::string_view name_str) {
-      tag_extracted_name_ = tls.tagProducer().produceTags(name_str, tags_);
-    });
+  StatNameTagHelper(ThreadLocalStoreImpl& tls, StatName name,
+                    const StatNameTagVector& stat_name_tags, SymbolTable& symbol_table,
```

Comment:
you could use `absl::optional<StatNameTagVector>` and drop the bool.

Also you can get symbol_table from `tls.symbolTable()` so no need to pass it separately.